### DISCUSSION TITLE
How to validate a union type of an enum and a collection of this enum in query parameters?

### DISCUSSION THREAD

**Discussion by gloomybrain** [TypedProject/ts-express-decorators#1734] [Q&A] [answered]
Hi there! So I have an enum type:
```typescript
enum Color {
  RED: 'red',
  GREEN: 'green',
  BLUE: 'blue'
}
```

And I have a model class that describes all possible query parameters for a REST request:
```typescript
@AdditionalProperties(false)
class GetColorsQueryParameters {
  @AnyOf(Color, array().items(Color))
  @OnDeserialize(deserializeColors)
  public color: Color | Color[]
}
```
What I want to achieve here is to allow the following requests:
 - `GET /?color=green`
 - `GET /?color[]=red&color[]=blue`

And I obviously want to forbid any other query parameters along with unsupported colors from being queried.

Please don't mind the `deserializeColors` because it's there just for the sake of demonstration. [It is clear](https://github.com/tsedio/tsed/discussions/1732) that a custom deserialization is required for union types to work.
However what is not quite clear to me is how do I go about writing a working _validation_ for this use-case? The `@AnyOf(Color, array().items(Color))` doesn't actually work: the validation pipeline passes incorrect values further down to `deserializeColors`, meaning that `ajv` schema is either not used at all or at least is ineffective in this case. Sending, say `GET /?color=purple` does not lead to a validation error.
So what would be the right way to build a working validation schema for such a union type?

P.S.
I've come across [this Stack Overflow question](https://github.com/tsedio/tsed/discussions/1732) recently but it seems to be not working for current version of Ts.ED.

**Comment by Romakita** [comment, marked as answer]
Because `@AnyOf(Color, array().items(Color))` is probably not correct.

Try that:

```typescript
  @CollectionOf(Color)
  @OnDeserialize(deserializeColors)
  public color: Color[]
```

I think it should works, because, the ValidationService will try to coerce data before validating it and the deserializeColors will map correctly the data to an array

See you
romain

**Reply by gloomybrain** [reply]
Hey @Romakita thanks for you answer. I've tried it on my codebase and it appears that the type coercion you've mentioned didn't work for the single parameter case.
The request looks like this: `GET /?color=red`
And the error looks like the following:
```
PARAM_VALIDATION_ERROR: Bad request on parameter "request.query"
    GetColorsQueryParameters.color should be object. Given value: "red"
        at Function.from (${project_root}/node_modules/@tsed/platform-params/src/errors/ParamValidationError.ts:19:21)
```

**Reply by Romakita** [reply]
hum... maybe the coerce type is missing for this case. I'll try to fix this issue

**Reply by gloomybrain** [reply]
That's truly awesome! However if you happen to know any workaround for this (I mean a union type of enum / array of enum in query parameters) please don't hesitate to share it with me

**Reply by Romakita** [reply]
Here the right example:
```typescript
enum Color {
      RED = "red",
      GREEN = "green",
      BLUE = "blue"
    }

    @AdditionalProperties(false)
    class GetColorsQueryParameters {
      @AnyOf(string().enum(Color), array().items(string().enum(Color)))
      public color: Color | Color[];
    }
```
This example will be valid with the next Ts.ED version.

Romain

**Reply by gloomybrain** [reply]
Huuuugee thanks to you @Romakita!
